File: 2023/day1/day1.rs
use std::fs::read_to_string;
use std::path::Path;

fn main() {
    //    read_file(true);

    let lines = read_lines(true);
    let mut v: Vec<i32> = Vec::new();
    let valid_digits: [&str; 10] = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"];

    for line in lines {
        let mut is_first = true;
        let mut first_number = char::default();
        let mut last_number = char::default();

        for elt in line.chars() {
            if elt.is_numeric() {
                if is_first {
                    first_number = elt;
                    is_first = false;
                } else {
                    last_number = elt;
                }
            }
        }


        let n_digits = count_digits(line.clone());

        let mut a_string = String::from("");
        a_string.push(first_number);
        if n_digits == 1 {
            a_string.push(first_number);
        } else {
            a_string.push(last_number);
        }
        let my_int: i32 = a_string.parse().unwrap();
        v.push(my_int);
    }
    let sum: i32 = v.iter().sum();
    println!("{}", sum)
}

fn count_digits(line: String) -> i32 {
    let mut counter: i32 = 0;
    for elt in line.chars() {
        if elt.is_numeric() {
            counter += 1;
        }
    }
    counter
}

fn read_lines(example: bool) -> Vec<String> {
    let file_path = if example {
        Path::new("./example_in.txt")
    } else {
        Path::new("./input.txt")
    };

    read_to_string(file_path)
        .unwrap() // panic on possible file-reading errors
        .lines() // split the string into an iterator of string slices
        .map(String::from) // make each slice into a string
        .collect() // gather them together into ,a vector
}

File: 2024/day_1/day1.py
import regex as re
import numpy as np

INPUT_FILE_PATH = './input.txt'
INPUT_FILE_PATH = './example_in.txt'
DIGIT_WORDS = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
RE = '\d|' + '|'.join(DIGIT_WORDS)

def main():
    lines = parse_input_file()
    print(lines)
    cols = [l.split("   ") for l in lines]
    cols = np.array(cols).T.astype(int)
    list_1 = list(cols[0])
    list_2 = list(cols[1])

    all_dists=[]
    while len(list_1)>0:
        argmini_1 = np.argmin(list_1)
        mini_1 = list_1.pop(argmini_1)
        
        argmini_2 = np.argmin(list_2)
        mini_2 = list_2.pop(argmini_2)
        dist = abs(mini_1-mini_2)
        all_dists.append(dist)
    print(sum(all_dists))

def parse_input_file():
    with open(INPUT_FILE_PATH, 'r') as f:
        lines = f.read().split("\n")
    return lines

if __name__ == "__main__":
    main()